components/script/dom/bindings/typedarrays.rs

```diff
@@ -68,6 +66,20 @@ where
         data
     }
 
+    /// <https://tc39.es/ecma262/#sec-detacharraybuffer>
+    pub fn detach_internal(&self, cx: JSContext) -> bool {
+        assert!(self.is_initialized());
+        let mut is_shared = false;
+        unsafe {
+            assert!(JS_IsArrayBufferViewObject(*self.internal.handle()));
```

Comment:
The assert is failing despite what the [documentation](https://github.com/servo/mozjs/blob/1d0ba1dc0c40008cfea6f14271844558dbea3cbd/mozjs-sys/mozjs/js/public/experimental/TypedData.h#L254) suggests:

```
/**
 * Return the ArrayBuffer or SharedArrayBuffer underlying an ArrayBufferView.
 * This may return a detached buffer.  |obj| must be an object that would
 * return true for JS_IsArrayBufferViewObject().
 */
 ```

Reply:
After removing assert I am getting this panic:

```
  ▶ CRASH [expected OK] /_webgpu/webgpu/cts.https.html?q=webgpu:api,operation,adapter,requestAdapter:requestAdapter:*
  │ 
  │ UNSUPPORTED (log once): POSSIBLE ISSUE: unit 1 GLD_TEXTURE_INDEX_2D is unloadable and bound to sampler type (Float) - using zero texture because texture unloadable
  │ Caught signal 11 in thread "Script(1,1)"
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::rt::Argument::fmt
  │              at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/fmt/rt.rs:138:9
  │       core::fmt::write
  │              at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/fmt/mod.rs:1114:21
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::crash_handler::install::handler
  │    5: __platform_memmove
  │    6: _ZL22AnnotateMozCrashReasonPKc
  │              at /Users/taym95/workspace/opensource/servo/target/production/build/mozjs_sys-eaa49f285018e146/out/build/dist/include/mozilla/Assertions.h:43:3
  │       _ZN8JSObject13maybeUnwrapAsIN2js21ArrayBufferViewObjectEEEPT_v
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/JSObject.h:645:3
  │       _Z27JS_GetArrayBufferViewBufferP9JSContextN2JS6HandleIP8JSObjectEEPb.cold.1
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/ArrayBufferViewObject.cpp:210:16
  │    7: _ZL22AnnotateMozCrashReasonPKc
  │       _ZN8JSObject13maybeUnwrapAsIN2js21ArrayBufferViewObjectEEEPT_v
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/JSObject.h:645:3
  │       _Z27JS_GetArrayBufferViewBufferP9JSContextN2JS6HandleIP8JSObjectEEPb
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/ArrayBufferViewObject.cpp:210:16
  │    8: script::dom::bindings::typedarrays::HeapTypedArray<T>::detach_internal
  │    9: <script::dom::gpubuffer::GPUBuffer as script::dom::bindings::codegen::Bindings::WebGPUBinding::GPUBuffer_Binding::GPUBufferMethods>::Unmap
  │   10: <script::dom::gpubuffer::GPUBuffer as script::dom::bindings::codegen::Bindings::WebGPUBinding::GPUBuffer_Binding::GPUBufferMethods>::Destroy
  │   11: script::dom::bindings::codegen::Bindings::WebGPUBinding::GPUBuffer_Binding::destroy::{{closure}}
  │   12: mozjs::panic::wrap_panic
  │   13: script::dom::bindings::codegen::Bindings::WebGPUBinding::GPUBuffer_Binding::destroy
  │   14: _CallJitMethodOp
  │   15: script::dom::bindings::utils::generic_method
  │   16: _Z12CallJSNativeP9JSContextPFbS0_jPN2JS5ValueEEN2js10CallReasonERKNS1_8CallArgsE
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/Interpreter.cpp:486:13
  │       _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
  │              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/vm/Interpreter.cpp:580:12
  │   17: _ZN2js3jit14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_14ICFallbackStubEjPN2JS5ValueENS7_13MutableHandleIS8_EE
  └              at /Users/taym95/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/7184f65/mozjs-sys/mozjs/js/src/jit/BaselineIC.cpp:1591:10
  ```

Reply:
Thanks for testing!

So, I think this is the problem:

1. previous code in `gpubuffer` would simply call `DetachArrayBuffer` on the object, whereas here we first do `JS_GetArrayBufferViewBuffer`(which matches what was done `acquire_data`). 
2. Not all internal objects are of the same type, because that created by `NewExternalArrayBuffer` is an array buffer, whereas the rest are typed arrays "backed" by one(I assume).

Solution: if `JS_IsArrayBufferViewObject` returns false, use `JS_GetArrayBufferViewBuffer` to get the buffer, else the internal is the buffer already. Then, call `DetachArrayBuffer` on the buffer. 

Let me know how it goes.

Reply:
Updated the boolean logic above... (buffer vs buffer view)